### DISCUSSION TITLE
Support saving environment variables between steps

### DISCUSSION THREAD

**Discussion by daandemeyer** [community/community#26407] [Actions]
<p>With <a href="https://github.community/t/yaml-folded-multiline-strings-interpreted-as-multiple-commands/16179">https://github.community/t5/GitHub-Actions/YAML-folded-multiline-strings-interpreted-as-multiple-commands/m-p/30876#M593</a> solved, the only thing standing in the way of reusing a lot of commands between Linux/MacOS and Windows when building C++ with CMake is an option to keep environment variables between steps (currently the environment is reset after each step).</p>
<p>For example, an option <code>transfer-env</code> that, when enabled, saves the environment after each step and restores it in the next step would allow me to remove a lot of duplicate code from reproc’s (<a href="https://github.com/DaanDeMeyer/reproc/blob/master/.github/workflows/main.yml" rel="nofollow noopener">https://github.com/DaanDeMeyer/reproc/blob/master/.github/workflows/main.yml</a>) and doctest’s WIP (<a href="https://github.com/DaanDeMeyer/doctest/blob/github-actions/.github/workflows/main.yml" rel="nofollow noopener">https://github.com/DaanDeMeyer/doctest/blob/github-actions/.github/workflows/main.yml</a>) Github Actions configs.</p>
<p>I’ve seen that it’s possible to set global environment variables with <code>exportVariable</code> in <a href="https://github.com/actions/toolkit" rel="nofollow noopener">https://github.com/actions/toolkit</a> which makes me think that adding this is possible. If possible I’d try to implement it myself but I do not think the relevant code is open source.</p>

**Comment by daandemeyer** [comment]
<p>Using set-env I was able to achieve the desired result.</p>
<p>On Windows, VS related environment variables are set by a third-party batch script, so I had to add them with set-env manually after calling the batch script. Instead of filtering the relevant environment variables, I just make every environment variable global using set-env. The script can be found <a href="https://github.com/DaanDeMeyer/reproc/blob/master/.github/workflows/vsenv.bat" rel="nofollow noopener">here</a>.</p>

**Comment by chrispat** [comment]
<p>Thanks for creating that script.  I would be a little careful with that script as you are stomping on PATH rather an updating it with append-path.  In most cases it probably won’t be an issue but you never know. </p>

**Comment by kevbite** [comment]
<p>I’ve been going round in cycles with this as I had my set-env like</p>
<p>::set-env VERSION=0.0.21-preview</p>
<p>Which looks like it’s working but it needs to be</p>
<p>::set-env name=VERSION::0.0.21-preview</p>

**Comment by iwdgo** [comment]
<p>Link has moved to:</p>
<p><a href="https://help.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-environment-variable" rel="nofollow noopener">https://help.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-environment-variable</a></p>

**Comment by hisuwh** [comment]
<p>How do I actually use these variables?</p>
<p>I’m setting a variable like this in a script:</p>
<pre><code class="lang-auto">echo "::set-env name=app_version::$version"
</code></pre>
<p>And trying to use it in my workflow:</p>
<pre><code class="lang-auto">    - name: Upload Artifacts
      uses: actions/upload-artifact@v2
      with:
        name: app-${{ app_version }}
        path: ${{ github.workspace }}/path-to-upload
</code></pre>
<p>But before my action runs I get this error:</p>
<pre><code class="lang-auto">Invalid workflow file
The workflow is not valid. .github/workflows/publish.yml (Line: 55, Col: 15): Unrecognized named-value: 'app_version'.
</code></pre>
<p>I have tried initialising it at both the job and workflow level like below but I still get the same error:</p>
<pre><code class="lang-auto">env:
   app_version: 0.0.0
</code></pre>

**Comment by airtower-luna** [comment]
<p>The <code>set-env</code> command sets an environment variable, so you have to access it via the <code>env</code> context:</p>
<pre><code class="lang-auto">      with:
        name: app-${{ env.app_version }}
</code></pre>
<p>When inside a <code>run</code> command or some called program use whatever means the programming language offers to access environment variables, e.g. <code>"${app_version}"</code> in bash.</p>

**Comment by awmatheson** [comment, marked as answer]
<p>This is being deprecated due to a vulnerability - see <a href="https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/" rel="noopener nofollow ugc">https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/</a></p>
<p>Now use:</p>
<pre><code class="lang-auto">echo "action_state=yellow" &gt;&gt; $GITHUB_ENV
</code></pre>

**Reply by amano-takahisa** [reply]
Link to official documents.
https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-environment-variable

**Comment by chrispat** [comment]
<p>This method was deprecated due to a security vulnerability please see <a href="https://github.community/t/support-saving-environment-variables-between-steps/16230/9" class="inline-onebox">Support saving environment variables between steps</a> for the new method.</p>

**Comment by kiptoomm** [comment]
<p>It’s also possible to use job-scoped environment variables, no? See this link: <a href="https://kevsoft.net/2020/04/25/creating-a-job-scoped-computed-environment-variable-in-github-actions.html" rel="noopener nofollow ugc">https://kevsoft.net/2020/04/25/creating-a-job-scoped-computed-environment-variable-in-github-actions.html</a></p>
<p>for example:</p>
<pre><code class="lang-auto">name: Example Workflow
on: [push]

jobs:
  example-job:
    name: An example job
    runs-on: ubuntu-latest     
    env:
      MY_COMMON_ENV: 'XYZ'
    steps:
        - name: step 1 
           run: echo $MY_COMMON_ENV
        - name: step 2
           run: echo $MY_COMMON_ENV
           ...
</code></pre>

**Comment by Simran-B** [comment]
<p>Yes, but if you need to set an environment variable dynamically in a step, then you need to use environment files.</p>
